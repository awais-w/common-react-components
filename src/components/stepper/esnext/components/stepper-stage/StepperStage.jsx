/**
 * Our <Stepper/> component handles drawing content / headers
 * This just provides a cleaner interface to construct <Stepper/>
 */
import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import Waypoint from 'react-waypoint';
import { Element } from 'react-scroll';
import uuidv4 from 'uuid/v4';
import bem from '@argos/utils/esnext/bem';
import classNames from 'classnames';
import scrollTo from '../../helpers/StepperFunctions';
import './StepperStage.scss';

const cn = bem({ prefix: 'cmc-', block: 'stepper-stage' });

class StepperStage extends PureComponent {
  uniqueId;

  constructor(props) {
    super(props);
    this.uniqueId = uuidv4();
  }

  scrollToMe() {
    scrollTo({
      componentName: this.uniqueId,
      duration: this.props.allowNavigation ? 500 : 0,
      parentID: this.props.parentScrollContainer,
    });
  }

  componentDidMount() {
    this.scrollToMe();
  }

  componentDidUpdate(prevProps) {
    if (
      prevProps.activeStepIndex !== this.props.activeStepIndex &&
      this.props.activeStepIndex === this.props.stepIndex
    ) {
      this.scrollToMe();
    }
  }

  render() {
    const {
      // Provided as part of <StepperStage/> props
      title,
      subTitle,
      content,
      // Added by <Stepper />
      stageType,
      stepIndex,
      onWaypointEnter,
      shouldNavBarUpdateOnScroll,
      shouldShowStageTitle,
      allowNavigation,
      onClickPreviousStep,
    } = this.props;

    const classes = classNames(cn(), {
      disabled: stageType !== 'current',
      locked: !allowNavigation,
    });

    return (
      <Fragment>
        <div
          key={`step-${stepIndex}`}
          onClick={
            allowNavigation && stageType === 'previous' ? () => onClickPreviousStep({ desiredIndex: stepIndex }) : null
          }
          className={classes}
          name={this.uniqueId}>
          <Element key={`step-${stepIndex}`} name={`step-${stepIndex}`}>
            {/* TODO - Provide prop */}
            {shouldShowStageTitle && (
              <div className={cn('headings')}>
                <h1>{title}</h1>
                {subTitle && <h4> - {subTitle}</h4>}
              </div>
            )}
            <div className={cn('content')}>{content}</div>
          </Element>
        </div>

        {shouldNavBarUpdateOnScroll && (
          <Waypoint
            onEnter={() => {
              onWaypointEnter({ stepIndex });
            }}
          />
        )}
      </Fragment>
    );
  }
}

StepperStage.propTypes = {
  /**
   * Primary title for this step
   */
  title: PropTypes.string.isRequired,
  /**
   * Show whats been completed in a previous step
   */
  subTitle: PropTypes.string,
  /**
   * Content for the current stage being rendered
   */
  content: PropTypes.node.isRequired,
  /**
   * Set the current step ( Zero based index )
   */
  activeStepIndex: PropTypes.number,
  /**
   * Where abouts this step is located in relation to current ( So we can for example )
   * Provide the ability to navigate back to stages marked as "Previous"
   */
  stageType: PropTypes.oneOf(['current', 'previous', 'later']),
  /**
   * The index of this stage in relation to other <StepperStage />'s
   */
  stepIndex: PropTypes.number,
  /**
   * When we scroll this stage into view, this parent function will be called to update NavBar
   */
  onWaypointEnter: PropTypes.func,
  /**
   * Used to update our parent with a new activeStepID when we click on previous stages
   */
  onClickPreviousStep: PropTypes.func,
  /**
   * Helper function in <Stepper/> to scroll to a component given it's container ID
   */
  shouldNavBarUpdateOnScroll: PropTypes.bool,
  /**
   * Should we show a title above this current stage content
   */
  shouldShowStageTitle: PropTypes.bool,
  /**
   * The parent container within which the scrolling needs to happen.
   */
  parentScrollContainer: PropTypes.string,
  /**
   * Flag to indicate whether navigation is allowed from stepper stage or not.
   */
  allowNavigation: PropTypes.bool,
};

StepperStage.defaultProps = {
  subTitle: '',
  // provide default values for those provided by <Stepper/>
  activeStepIndex: 0,
  stepIndex: 0,
  stageType: 'current',
  onWaypointEnter: () => {},
  onClickPreviousStep: () => {},
  scrollTo: () => {},
  shouldNavBarUpdateOnScroll: true,
  shouldShowStageTitle: false,
  parentScrollContainer: '',
};

export default StepperStage;
