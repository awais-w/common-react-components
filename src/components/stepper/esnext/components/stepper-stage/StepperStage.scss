@import '@cmc/common-styles/src/scss/_variables.scss';

.cmc-stepper-stage {
  @include font-standard;
  min-height: 100%;
  padding: $double;
  &__headings {
    display: inline-block;

    h1,
    h4 {
      display: inherit;
    }
  }
  &__page-divider {
    margin-bottom: $large;
  }
}
