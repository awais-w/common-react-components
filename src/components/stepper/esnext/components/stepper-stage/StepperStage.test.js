import React from 'react';
import Waypoint from 'react-waypoint';
import { shallow } from 'enzyme';
import StepperStage from './StepperStage';

const previousProps = {
  title: 'title',
  content: 'content',
  // renderUpToLatestStageReached: false,
  activeStepIndex: 0,
  stepIndex: 1,
  stageType: 'previous',
  allowNavigation: true,
  shouldNavBarUpdateOnScroll: true,
  onClickPreviousStep: jest.fn(),
};

const currentProps = {
  ...previousProps,
  activeStepIndex: 1,
  stageType: 'current',
  shouldNavBarUpdateOnScroll: false,
};

// Mock the uuid
jest.mock('uuid/v4', () => () => '00000000-0000-0000-0000-000000000000');

const setup = ({ props }) => {
  const component = shallow(<StepperStage {...props} />);
  return {
    wrapper: component,
    getStepperStage: () => component.find('.cmc-stepper-stage'),
    getWaypoint: () => component.find(Waypoint),
    getHeadings: () => component.find('.cmc-stepper-stage__headings'),
    getHeadingsSubtitle: () => component.find('.cmc-stepper-stage__headings').find('h4'),
  };
};

const stepperFunctions = require('../../helpers/StepperFunctions');

stepperFunctions.default = jest.fn();

describe('StepperStage', () => {
  describe('base tests', () => {
    const component = setup({ props: previousProps });
    it('exists', () => {
      expect(component.wrapper).toExist();
    });
    it('should match default snapshot', () => {
      expect(component.wrapper).toMatchSnapshot();
    });
    it('calls default functions successfully', () => {
      component.wrapper.instance().props.onWaypointEnter();
      component.wrapper.instance().props.onClickPreviousStep();
      component.wrapper.instance().props.scrollTo();
    });
  });

  describe('when scrollToMe is called', () => {
    const component = setup({ props: previousProps });
    component.wrapper.instance().scrollToMe();
    it('should call scrollTo', () => {
      expect(stepperFunctions.default).toBeCalled();
    });
  });

  describe('when componentDidMount is called', () => {
    const component = setup({ props: previousProps });
    component.wrapper.instance().scrollToMe = jest.fn();
    component.wrapper.instance().componentDidMount();
    it('should call scrollToMe', () => {
      expect(component.wrapper.instance().scrollToMe).toBeCalled();
    });
  });

  describe('headings div', () => {
    const component = setup({ props: previousProps });
    it('should not render headings if shouldShowPageTitle is set to false', () => {
      component.wrapper.setProps({ shouldShowStageTitle: false });
      expect(component.getHeadings()).not.toExist();
    });
    it('should render headings with no subtitle if shouldShowPageTitle is set to true and subtitle is null', () => {
      component.wrapper.setProps({ shouldShowStageTitle: true });
      expect(component.getHeadings()).toExist();
      expect(component.getHeadingsSubtitle()).not.toExist();
    });
    it('renders headings with subtitle if shouldShowPageTitle is set to true and subtitle is not null', () => {
      component.wrapper.setProps({ shouldShowStageTitle: true, subTitle: 'Test' });
      expect(component.getHeadingsSubtitle().text()).toEqual(` - ${component.wrapper.instance().props.subTitle}`);
    });
  });

  describe('when componentDidUpdate is called', () => {
    it('calls scrollToMe if activeStepIndex changes and is now equal to stepIndex', () => {
      const component = setup({ props: previousProps });
      component.wrapper.instance().scrollToMe = jest.fn();
      component.wrapper.setProps(currentProps);
      expect(component.wrapper.instance().scrollToMe).toBeCalled();
    });
    it('does not call scrollToMe if activeStepIndex does not change', () => {
      const component = setup({ props: previousProps });
      component.wrapper.instance().scrollToMe = jest.fn();
      component.wrapper.setProps(previousProps);
      expect(component.wrapper.instance().scrollToMe).not.toBeCalled();
    });
  });

  describe('onWaypointEnter', () => {
    it('should be called when onEnter event is fired with shouldNavBarUpdateOnScroll set to true', () => {
      const component = setup({ props: { ...previousProps, onWaypointEnter: jest.fn() } });
      component.getWaypoint().simulate('enter');
      expect(component.wrapper.instance().props.onWaypointEnter).toBeCalledWith({ stepIndex: previousProps.stepIndex });
    });
  });

  describe('when is the previous stage', () => {
    let component = setup({ props: previousProps });
    it('is disabled', () => {
      expect(component.getStepperStage()).toHaveClassName('disabled');
    });
    describe('the navigation to previous stages', () => {
      it("is not allowed when it's locked", () => {
        component = setup({
          props: { ...previousProps, allowNavigation: false, onClickPreviousStep: jest.fn() },
        });
        expect(component.getStepperStage()).toHaveClassName('locked');
        expect(component.wrapper.instance().props.onClickPreviousStep).not.toBeCalled();
        component.getStepperStage().simulate('click');
        expect(component.wrapper.instance().props.onClickPreviousStep).not.toBeCalled();
      });
      it("is allowed when it's NOT locked", () => {
        component = setup({ props: { ...previousProps, allowNavigation: true, onClickPreviousStep: jest.fn() } });
        expect(component.getStepperStage()).not.toHaveClassName('locked');
        expect(component.wrapper.instance().props.onClickPreviousStep).not.toBeCalled();
        component.getStepperStage().simulate('click');
        expect(component.wrapper.instance().props.onClickPreviousStep).toBeCalledWith({
          desiredIndex: previousProps.stepIndex,
        });
      });
    });
  });
  describe('when is the current stage', () => {
    const component = setup({ props: currentProps });
    it('is not disabled', () => {
      expect(component.getStepperStage()).not.toHaveClassName('disabled');
    });
    describe('the navigation to previous stages', () => {
      describe('is not allowed', () => {
        expect(component.getStepperStage()).not.toHaveClassName('locked');
        expect(component.wrapper.instance().props.onClickPreviousStep).not.toBeCalled();
        component.getStepperStage().simulate('click');
        expect(component.wrapper.instance().props.onClickPreviousStep).not.toBeCalled();
      });
    });
  });
});
