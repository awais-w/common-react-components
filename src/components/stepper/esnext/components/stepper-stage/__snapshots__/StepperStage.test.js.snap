// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepperStage base tests should match default snapshot 1`] = `
<Fragment>
  <div
    className="cmc-stepper-stage disabled"
    key="step-1"
    name="00000000-0000-0000-0000-000000000000"
    onClick={[Function]}
  >
    <Element
      key="step-1"
      name="step-1"
    >
      <div
        className="cmc-stepper-stage__content"
      >
        content
      </div>
    </Element>
  </div>
  <Waypoint
    bottomOffset="0px"
    fireOnRapidScroll={true}
    horizontal={false}
    onEnter={[Function]}
    onLeave={[Function]}
    onPositionChange={[Function]}
    topOffset="0px"
  />
</Fragment>
`;
